### PR TITLE
Cache global.json in NuGet SDK resolver

### LINKED ISSUE #11450: [DCR]: NuGet SDK resolver should cache global.json itself
### NuGet Product(s) Affected

MSBuild.exe, dotnet.exe

### Current Behavior

Currently the NuGet SDK resolver relies on the MSBuild SDK resolver APIs to save state.  However, the state is not guaranteed to be thread-safe and so currently the NuGet SDK resolver is parsing global.json many times by Visual Studio.  This is because the state is only set after a resolver completes and the cache key is a build submission.  Visual Studio issues several design time builds will load global.json over and over.

### Desired Behavior

The NuGet SDK resolver should have its own file cache that only loads a file once unless it has changed.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jeffkl**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11450

Regression? No

## Description
The MSBuild SDK resolver APIs provide some state but it is not thread-safe.  This results in the NuGet SDK resolver loading and parsing global.json N times in Visual Studio, once for solution load and once for every design time build.

This change introduces a file cache based on path and last write time that is thread-safe.

I also changed how global.json is parsed to help speed up the process.

Related to https://github.com/NuGet/Home/issues/11443 and https://github.com/NuGet/Home/issues/11441

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (jeffkl) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: I ended up using a custom file cache instead of [System.Runtime.Caching.MemoryCache](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.caching.memorycache) because that would require me to pull in a new assembly for .NET Framework and ship it next to our assemblies in Visual Studio.

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs
Comment (jeffkl) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs]: This is the most meaningful part of the change.  Rather than rely on the state from MSBuild which is not thread-safe, we now parse versions from `global.json` and cache the result based on the file path and last write time.  This means the file is only parsed once per instance so Visual Studio won't read it over and over.

**Comment by jeffkl** [comment]
Note to self, do not merge until dev targets 17.2

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (erdembayar) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: Could we use remove func here? Because previously removing func for concurrent dictionary ( https://github.com/NuGet/NuGet.Client/pull/4095/files ) improved perf, I believe this is very similar situation.

Comment (erdembayar) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: Also do we really need `Lazy` here? Because we always end up evaluating the value.

**Review by erdembayar — COMMENTED** [review]
LGTM. Just 1 comment.

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (jeffkl) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: Great question.  Yes the `Lazy<T>` is necessary because of how `ConcurrentDictionary<TKey, TValue>.AddOrUpdate()` works:

[From the docs: ](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-6.0#System_Collections_Concurrent_ConcurrentDictionary_2_AddOrUpdate__0__1_System_Func__0__1__1__)

> For modifications and write operations to the dictionary, ConcurrentDictionary<TKey,TValue> uses fine-grained locking to ensure thread safety. (Read operations on the dictionary are performed in a lock-free manner.) The addValueFactory and updateValueFactory delegates may be executed multiple times to verify the value was added or updated as expected. However, they are called outside the locks to avoid the problems that can arise from executing unknown code under a lock. Therefore, AddOrUpdate is not atomic with regards to all other operations on the ConcurrentDictionary<TKey,TValue> class.

This means that two threads will only add one item, but end up calling the factory method if they're both called at the same time.  To ensure that the factory is only called once, you wrap it in a `Lazy<T>`.  The `Lazy` object can be created multiple times, but all of the calls receive the object that was created first.  All threads then access `Value` which is thread-safe so all threads wait for the factory to actually run (if it hasn't already) or they receive the cached value.

One blog post that explains more is here: https://andrewlock.net/making-getoradd-on-concurrentdictionary-thread-safe-using-lazy/

I've been bitten by this in the past which is the only reason I know about this limitation: https://github.com/dotnet/msbuild/pull/4797

**Review by nkolev92 — APPROVED** [review]

**Review by erdembayar — APPROVED** [review]

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (zivkan) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: As far as I can tell, `wasRead` and `wasGlobalJsonRead` are only used for testing. Which is ok, except that MSBuild SDK resolving is on the hot path for project load (both in VS and command line).  The reason I bring up `wasRead` is because it's captured scope from the `GetMSBuildSdkVersions` method, and anonymous methods cause allocations. In this case, there's one anonymous method passed to `Lazy<T>`, and allocating the `Lazy<T>` itself is inside another anonymous method that is passed to `ConcurrentDictionary.AddOrUpdate`.

If `wasRead` is removed, then I believe that these methods could then be static and moved to class methods, allowing the delegates passed to `GetOrAdd` and `Lazy<T>`'s constrctor to be zero allocation.

I'll understand if some people consider the perf impact acceptable to validate correctness. I just wanted to make sure this was considered since `wasRead`/`wasGlobalJsonRead` is not used in production code.

Comment (jeffkl) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: This is a good point.  Adding testability at the sacrifice of performance here probably isn't great.  What do you think about including this only for Debug bits?  For example:

```c#
#if DEBUG
   wasRead = true;
#endif
```

I'm not sure if our builds that run tests only build Debug.  I could `ifdef` the tests out on Release bits I suppose.  What does everybody think?

Comment (zivkan) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: Even if you don't capture scope, simply using an anonymous method will still cause allocations (unless the C# compiler has improvements I'm not aware of). So, using `#if` to stop using captured scope, but still using the anonymous method, then we won't get any perf benefit.

I believe the only way to remove allocations when passing delegates is to use a named class method (I hope I'm wrong)

Comment (jeffkl) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: I've decided to use an interface to make the class more testable and used an event instead which fires if global.json is read.  Let me know what you think!

**Comment by jeffkl** [comment]
@zivkan do you want to review this again before I merge?

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Strings.resx
Comment (zivkan) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Strings.resx]: There need to be comments telling translators what `{0}` and {1}` mean

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Strings.resx
Comment (zivkan) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Strings.resx]: There need to be comments telling translators what `{0}` and {1}` mean

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (dtivel) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: This looks like non-standard usage passing an event argument for `sender`.  How about this?
```C#
public event EventHandler<string> FileRead;

...

fileReadEventHandler?.Invoke(this, filePath);
```

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (dtivel) [inline src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs:138]: If I understand correctly, before this change, it was assured that if the `msbuild-sdks` property was read successfully that the entire global.json was well-formed; however, with this change, this assurance no longer exists.  For example, the top-level JSON object could be lacking a closing `}`.  I presume that such an error would be now caught by some later reader.

Is this a concern?

**Review by dtivel — APPROVED** [review]

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Strings.resx
Comment (jeffkl) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/Strings.resx]: Thanks, I'll add comments for all of these!

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (jeffkl) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs]: Thanks, I was looking for something like this!

File: src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs
Comment (jeffkl) [reply src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs:138]: Yes the parsing is _slightly_ more permissive at the cost of being faster.  Invalid JSON before and after the `msbuild-sdks` section will still produce an error since I'm calling `.Skip()` to read all of the top-level elements.  Invalid **values** in the msbuild-sdks section are ignored now but invalid JSON is not.  In my opinion, its okay for this parsing logic to be less strict and I have added unit tests to verify what is and isn't going to give an error.

**Review by zivkan — APPROVED** [review]

**Pull request merged**
